File: java/amp-eventing/src/main/java/amp/eventing/RpcFilter.java
package amp.eventing;


import java.util.ArrayList;
import java.util.List;
import java.util.Timer;
import java.util.TimerTask;
import java.util.UUID;

import org.joda.time.Duration;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import amp.eventing.EnvelopeHelper;
import amp.eventing.EventContext.Directions;


public class RpcFilter implements IEventProcessor {

    protected static final Logger log = LoggerFactory.getLogger(RpcFilter.class);
    protected List<UUID> sentRequests;

    
    public RpcFilter() {
        sentRequests = new ArrayList<UUID>();
    }

    
    @Override
    public void processEvent(EventContext context, IContinuationCallback continuation) throws Exception {
    	
    	if (Directions.In == context.getDirection()) {
    		this.processInbound(context, continuation);
    	}
    	else if (Directions.Out == context.getDirection()) {
    		this.processOutbound(context, continuation);
    	}
    }
    
    
    public void processInbound(EventContext context, IContinuationCallback continuation) throws Exception {

        boolean ourOwnRequest = false;
        EnvelopeHelper env = new EnvelopeHelper(context.getEnvelope());

        try {
            if (env.IsRequest()) {
                UUID requestId = env.getMessageId();

                synchronized (this) {
                    if (sentRequests.contains(requestId)) {
                        log.info("Filtering out our own request: " + requestId.toString());
                        ourOwnRequest = true;
                    }
                }
            }
        } catch (Exception ex) {
            log.error("Failed to inspect an incoming event for potential filtering", ex);
        }

        if (!ourOwnRequest) { continuation.continueProcessing(); }
    }

    public void processOutbound(EventContext context, IContinuationCallback continuation) throws Exception {

        EnvelopeHelper env = new EnvelopeHelper(context.getEnvelope());

        if (env.IsRequest()) {
            final UUID requestId = env.getMessageId();
            Duration timeout = env.getRpcTimeout();

            synchronized (this) {
                log.debug(String.format("Adding requestId %s to the RPC Filter list", requestId.toString()));
                sentRequests.add(requestId);
            }

            if (timeout == Duration.ZERO) {
                Timer gc = new Timer(true);
                gc.schedule(new TimerTask() {

                    @Override
                    public void run() {
                        requestTimeout_GarbageCollect(requestId);
                    }

                }, timeout.getMillis() * 2);
            } else {
                log.warn(String.format(
                                "Request %s was sent without a timeout: it will never be removed from the RPC Filter list",
                                requestId.toString()));
            }
        }
        
        continuation.continueProcessing();
    }

    public void requestTimeout_GarbageCollect(UUID requestId) {
        synchronized (this) {
            log.debug(String.format("Removing requestId %s from the RPC Filter list", requestId.toString()));
            sentRequests.remove(requestId);
        }
    }
}

File: dotnet/source/amp.eventing/RpcFilter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;

using Common.Logging;

using cmf.bus;
using amp.bus;

namespace amp.eventing
{
    /// <summary>
    /// When you send a request of a type that you also subscribe to, it may be
    /// desirable to not receive your own requests.  This processor filters out
    /// requests that you have sent so that you do not receive them yourself.
    /// </summary>
    public class RpcFilter : IEventProcessor
    {
        protected IList<Guid> _sentRequests;
        protected object _listLock = new object();
        protected ILog _log;


        public RpcFilter()
        {
            _sentRequests = new List<Guid>();

            _log = LogManager.GetLogger(this.GetType());
        }


        public void ProcessEvent(EventContext context, Action continueProcessing)
        {
            if (EventContext.Directions.In == context.Direction)
            {
                this.ProcessInbound(context, continueProcessing);
            }
            if (EventContext.Directions.Out == context.Direction)
            {
                this.ProcessOutbound(context, continueProcessing);
            }
        }

        public void ProcessInbound(EventContext context, Action continueProcessing)
        {
            bool ourOwnRequest = false;
            Envelope env = context.Envelope;

            try
            {
                if (env.IsRequest())
                {
                    Guid requestId = env.GetMessageId();

                    lock (_listLock)
                    {
                        if (_sentRequests.Contains(requestId))
                        {
                            _log.Info("Filtering out our own request: " + requestId.ToString());
                            ourOwnRequest = true;
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                _log.Error("Failed to inspect an incoming event for potential filtering", ex);
            }

            if (!ourOwnRequest) { continueProcessing(); }
        }

        public void ProcessOutbound(EventContext context, Action continueProcessing)
        {
            Envelope env = context.Envelope;

            if (env.IsRequest())
            {
                Guid requestId = env.GetMessageId();
                TimeSpan timeout = env.GetRpcTimeout();

                lock (_listLock)
                {
                    _log.Debug(string.Format("Adding requestId {0} to the RPC Filter list", requestId.ToString()));
                    _sentRequests.Add(requestId);
                }

                if (timeout.IsNotEmpty())
                {
                    Timer gc = new Timer(this.RequestTimeout_GarbageCollect, requestId, timeout.Add(timeout), TimeSpan.Zero);
                }
                else
                {
                    _log.Warn(string.Format(
                        "Request {0} was sent without a timeout: it will never be removed from the RPC Filter list", 
                        requestId.ToString()));
                }
            }

            continueProcessing();
        }

        public void RequestTimeout_GarbageCollect(object requestGuid)
        {
            Guid requestId = (Guid)requestGuid;

            lock (_listLock)
            {
                _log.Debug(string.Format("Removing requestId {0} from the RPC Filter list", requestId.ToString()));
                _sentRequests.Remove(requestId);
            }
        }
    }




    public static class TimeSpanExtensions
    {
        public static bool IsNotEmpty(this TimeSpan timeout)
        {
            return !TimeSpan.Zero.Equals(timeout);
        }
    }
}
